hudi-common/src/main/java/org/apache/hudi/common/table/timeline/TimeGeneratorBase.java

```diff
@@ -85,20 +85,26 @@ public TimeGeneratorBase(HoodieTimeGeneratorConfig config, StorageConfiguration<
 
   protected LockProvider<?> getLockProvider() {
     // Perform lazy initialization of lock provider only if needed
-    if (lockProvider == null) {
+    String lockProviderClass = lockConfiguration.getConfig().getString("hoodie.write.lock.provider");
+    LOG.info("LockProvider for TimeGenerator: {}", lockProviderClass);
+    if (!LockProvider.isThreadSafe(lockProviderClass)) {
+      return createLockProvider(lockProviderClass);
```

Comment:
1. Please add an inline comment to explain the rationale behind returning a new instance for non-thread safe providers.
2. Creating a new instance for non-thread safe providers on every call might have performance and heap usage implications. Given that time generation has to happen on every start/end of operation, could this lead to poor driver memory utilization especially if multiple streams are multiplexed on the same driver?

Reply:
Can we just make the `LockProvider#tryLock` synchronized to make all of them thread-safe?

Reply:
Yes, it can cause issues with performance and heap usage. I think a better fix would be make it thread safe instead. We will not be able to make HiveMetastoreBasedLockProvider as thread safe though. So we would need a fix for HiveMetastoreBasedLockProvider atleast.

Reply:
Can we just make `TimeGenerator#generateTime` synchronized to fix the issue?

Reply:
Yeah. That should also work. Only issue could be contention when multiple streams are multiplexed on same driver. But it should be better than having a separate lock provider I think.

Reply:
I have pushed a change to make time generation API in SkewAdjustingTimeGenerator synchronized.
There is another issue around closing of lock provider. Currently lock provider is closed during unlock and if we are sharing the lock provider amongst multiple threads, we would need to fix the close semantics for lock provider. Will add it in HUDI-9254.

Reply:
> we would need to fix the close semantics for lock provider

+1.